[quant] Add reduce_range params for quantized_lstm

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39666 [quant] Enable per-channel quantization for LSTM Modules
* **#39604 [quant] Add reduce_range params for quantized_lstm**

Summary:
This change preserves BC for older models that are saved with reduce_range set to false.
Newer models will use the version information in RNN module to toggle reduce_range parameter

Internally this is implemented using a new CellParams type that calls the linear functions with reduce_range option set to true.
New models serialized will use the CellParams struct for the `__getstate__` and `__setstate__` calls. Older models using QuantizedCellParamsDynamic will continue to use their original serialization/de-serialization methods

tested using LSTM BC test and test_quantized_rnn

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21977600](https://our.internmc.facebook.com/intern/diff/D21977600)

supriyar committed:
[quant] Add reduce_range params for quantized_lstm
Summary:
This change preserves BC for older models that are saved with reduce_range set to false.
Newer models will use the version information in RNN module to toggle reduce_range parameter

Internally this is implemented using a new CellParams type that calls the linear functions with reduce_range option set to true.
New models serialized will use the CellParams struct for the `__getstate__` and `__setstate__` calls. Older models using QuantizedCellParamsDynamic will continue to use their original serialization/de-serialization methods

tested using LSTM BC test and test_quantized_rnn

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add reduce_range params for quantized_lstm"


Summary:
This change preserves BC for older models that are saved with reduce_range set to false.
Newer models will use the version information in RNN module to toggle reduce_range parameter

Internally this is implemented using a new CellParams type that calls the linear functions with reduce_range option set to true.
New models serialized will use the CellParams struct for the `__getstate__` and `__setstate__` calls. Older models using QuantizedCellParamsDynamic will continue to use their original serialization/de-serialization methods

tested using LSTM BC test and test_quantized_rnn

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -360,15 +365,65 @@ struct QuantizedCellParamsDynamic : public CellParamsBase {
+  }
+};
+
+struct QuantizedCellParamsDynamicReduceRange : public QuantizedCellParamsDynamic {
```
Do we still need quantizedCellParamsDynamic? Why not replace that with this struct?

supriyar replied:
Older models are serialized with QuantizedCellParamsDynamic so we need to keep it around for BC.

raghuramank100 replied:
Dont follow, this is not an aten op, so why does this affect serialization?

raghuramank100 commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -360,15 +365,65 @@ struct QuantizedCellParamsDynamic : public CellParamsBase {
+          _packed_w_hh, /* Prepacked Weight Tensor */
+      Tensor _b_ih, /* float Bias Tensor */
+      Tensor _b_hh /* float Bias Tensor */,
+      bool reduce_range)
```
reduce_range argument is not used here.

raghuramank100 commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -192,6 +200,13 @@ def permute_hidden(self, hx, permutation):
             return hx
         return apply_permutation(hx, permutation)
 
+    def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
```
We need to ensure that the test_backward_compatibility tests check for versioning for serialization. Can you please add a test for that?

supriyar replied:
Sure, I can do that in a follow up PR. Adding serialized `.pt` files doesn't work with the ghimport flow so I'll have to create a different PR for it.

raghuramank100 commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -360,15 +365,65 @@ struct QuantizedCellParamsDynamic : public CellParamsBase {
     at::Tensor bias_ih,
-    at::Tensor bias_hh) {
+    at::Tensor bias_hh,
+    bool reduce_range) {
```
Where is the default value for reduce_range set?

raghuramank100 reviewed (changes requested):
Please see comments, thanks

supriyar committed:
Update on "[quant] Add reduce_range params for quantized_lstm"


Summary:
This change preserves BC for older models that are saved with reduce_range set to false.
Newer models will use the version information in RNN module to toggle reduce_range parameter

Internally this is implemented using a new CellParams type that calls the linear functions with reduce_range option set to true.
New models serialized will use the CellParams struct for the `__getstate__` and `__setstate__` calls. Older models using QuantizedCellParamsDynamic will continue to use their original serialization/de-serialization methods

tested using LSTM BC test and test_quantized_rnn

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -340,8 +344,9 @@ struct QuantizedCellParamsDynamic : public CellParamsBase {
     std::vector<c10::intrusive_ptr<LinearPackedParamsBase>>
         packed_params_to_serialize{packed_w_ih, packed_w_hh};
 
+    string op_name = reduce_range_ ? "quantized_dynamic_bc" : "quantized_dynamic";
```
Why call this quantized_dynamic_bc? This is the newer version that has reduce_range=True.

raghuramank100 commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -360,15 +365,63 @@ struct QuantizedCellParamsDynamic : public CellParamsBase {
+        /*w_hh_packed=*/std::move(packed_params[1]),
+        /*bias_ih=*/std::move(tensors[0]),
+        /*bias_hh=*/std::move(tensors[1]),
+        /*reduce_range=*/true);
```
cc @jamesr66a : Is there a better way to handle versioning for serializing custom classes?

jamesr66a replied:
How about, instead of creating a new cell type, add the bool value for `reduce_range` in the `vector<int64_t>` field of `CellParamsSerializationType`. That is currently unused, but we can define a convention for `QuantizedCellParamsDynamic` that the 0-th element (if present) is the bool value for `reduce_range`. If there are no elements in that vector, assume `false`

This we we don't need to create an entirely new cell type and increase binary size

raghuramank100 replied:
@supriyar proposed this option yesterday, it feels a bit hacky as we are using this field like the version number. Is there a way to define a version field (but still keep BC) in CellParamsSerializationType?

supriyar committed:
Update on "[quant] Add reduce_range params for quantized_lstm"


Summary:
This change preserves BC for older models that are saved with reduce_range set to false.
Newer models will use the version information in RNN module to toggle reduce_range parameter

Internally this is implemented using a new CellParams type that calls the linear functions with reduce_range option set to true.
New models serialized will use the CellParams struct for the `__getstate__` and `__setstate__` calls. Older models using QuantizedCellParamsDynamic will continue to use their original serialization/de-serialization methods

tested using LSTM BC test and test_quantized_rnn

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
